File: insertion_sort/insertion_sort.go
package main

import "fmt"

func insertion_sort(arr []int) {
	for i := 0; i < len(arr); i++ {
		var j = 0
		for arr[j] < arr[i] && j < i {
			j++
		}
		for j <= i {
			arr[j], arr[i] = arr[i], arr[j]
			j++
		}
	}
}

func main() {
	var arr = []int{5, 4, 3, 2, 1}
	insertion_sort(arr)
	fmt.Println(arr)
}

File: insertion_sort/InsertionSort.java
public class InsertionSort {
    public static void main(String[] args) {
        int[] arr = { 5, 4, 3, 2, 1 };
        insertionSort(arr);
        System.out.println(java.util.Arrays.toString(arr));
    }

    public static void insertionSort(int[] arr) {
        int n = arr.length;
        for (int i = 1; i < n; i++) {
            int j = i;
            while (j > 0 && arr[j - 1] > arr[j]) {
                swap(arr, j - 1, j);
                j--;
            }
        }
    }

    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}
